rust/tools/authorization-logic/src/souffle/souffle_interface.rs

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2021 The Raksha Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/// This file contains code for interfacing with souffle (a dialect and
+/// implementation of Datalog).
+use std::fs;
+use std::process::Command;
+
+use crate::{
+    ast::*,
+    parsing::astconstructionvisitor,
+    souffle::{lowering_ast_datalog::*, souffle_emitter::*},
+};
+
+/// Given a filename containing policy code `emit_souffle` parses it,
+/// constructs an AST, translates the AST to the datalogIR, and emits a string
+/// contianing souffle code.
+pub fn emit_souffle(filename: &str) {
+    let source = fs::read_to_string(filename).expect("failed to read input in emit_souffle");
+    let prog = astconstructionvisitor::parse_program(&source[..]);
+    let dlir_prog = LoweringToDatalogPass::lower(&prog);
+    let souffle_code = SouffleEmitter::emit_program(&dlir_prog);
+    println!("souffle code from {}, \n{}", &filename, souffle_code);
+}
+
+/// Given a filename containing policy code `input_to_souffle_file` parses it,
+/// translates it to datalogIR, and emits souffle code to a new file with the
+/// same name as the input but with the .dl extension in the out_dir
+pub fn input_to_souffle_file(filename: &str, in_dir: &str, out_dir: &str) {
+    let source = fs::read_to_string(&format!("{}/{}", in_dir, filename))
+        .expect("failed to read input in input_to_souffle_file");
+    let prog = astconstructionvisitor::parse_program(&source[..]);
+    let dlir_prog = LoweringToDatalogPass::lower(&prog);
+    let souffle_code = SouffleEmitter::emit_program(&dlir_prog);
+    fs::write(&format!("{}/{}.dl", out_dir, filename), souffle_code)
+        .expect("failed to write output to file");
+}
+
+/// Given a parsed AstProgram, `ast_to_souffle_file` emits souffle code to a file.
+/// This is needed in addition to the above function since there are other
+/// passes that work on the highest level IR besides the one used to emit the
+/// main code.
```

Comment:
This applies to all other comments (and can be done separately in a different PR), but the usual wording for doc comments would be

```
/// Emits souffle code to a file....
```

(third person, without repeating the function name)